File: backend/core/layers.py
import numpy as np
from .activations import get_activation
from .initializers import get_initializer

class DenseLayer:
    def __init__(self, input_size, output_size, activation, weight_init):
        self.input_size = input_size
        self.output_size = output_size
        
        self.activation = activation
        self.activation_fn, self.activation_derivative_fn = get_activation(activation)

        init_fn = get_initializer(weight_init)
        self.W = init_fn(shape=(input_size, output_size))
        self.b = init_fn(shape=(output_size,))

        self._input_cache = None
        self._z_cache = None

    def forward(self, x):
        self._input_cache = x
        self._z_cache = x @ self.W + self.b
        return self.activation_fn(self._z_cache)

    def backward(self, dz):
        dW = self._input_cache.T @ dz
        db = np.sum(dz, axis=0)
        dx = dz @ self.W.T
        return dW, db, dx

    def apply_grad(self, dW, db, update_fn):
        self.W, self.b = update_fn(self.W, self.b, dW, db)

File: netlab/src/engine/layers.ts
import { getActivation } from './activations'
import { getInitializer } from './initializers'
import { addBias, matmul, sumAxis0, transpose } from './matrix'
import type { ActivationFn, WeightInit } from '../types'

export class DenseLayer {
    inputSize: number
    outputSize: number
    activation: ActivationFn
    activationFn: (z: number[][]) => number[][]
    activationDerivativeFn: ((z: number[][]) => number[][]) | null

    W: number[][]
    b: number[]

    private inputCache: number[][] | null = null
    private zCache: number[][] | null = null

    constructor(inputSize: number, outputSize: number, activation: ActivationFn, weightInit: WeightInit) {
        this.inputSize = inputSize
        this.outputSize = outputSize
        this.activation = activation

        const [fn, derivativeFn] = getActivation(activation)
        this.activationFn = fn
        this.activationDerivativeFn = derivativeFn

        const init = getInitializer(weightInit)
        this.W = init(inputSize, outputSize)
        this.b = init(1, outputSize)[0]
    }

    forward(x: number[][]): number[][] {
        this.inputCache = x
        this.zCache = addBias(matmul(x, this.W), this.b)
        return this.activationFn(this.zCache)
    }

    get lastZ(): number[][] {
        if (!this.zCache) throw new Error('forward() must be called before backward()')
        return this.zCache
    }

    backward(dz: number[][]): { dW: number[][]; db: number[]; dx: number[][] } {
        if (!this.inputCache) throw new Error('forward() must be called before backward()')
        const dW = matmul(transpose(this.inputCache), dz)
        const db = sumAxis0(dz)
        const dx = matmul(dz, transpose(this.W))
        return { dW, db, dx }
    }

    applyGrad(dW: number[][], db: number[], updateFn: (W: number[][], b: number[], dW: number[][], db: number[]) => [number[][], number[]]): void {
        const [W, b] = updateFn(this.W, this.b, dW, db)
        this.W = W
        this.b = b
    }
}
